### PR TITLE
feat(manager): register user with email

### PR DESCRIPTION
**Pull request by abtinokhovat**
- implementing JWT
- implementing login and register
	- handler	 
	- validator	
	- service
- adding richerror

### CODE REVIEW

**Review by Mdhesari — COMMENTED** [review]
One subject, if services are decoupled I think better manager request the create user process from source team otherwise everything is fine

**Comment by abtinokhovat** [comment]
> One subject, if services are decoupled I think better manager request the create user process from source team otherwise everything is fine

Do you mean transferring the auth logic to the source? I do not get your point please explain more

**Comment by Mdhesari** [comment]
I meant touching the delivery folder for source service, I think it would be better to have a good structure in order to PR to services, for example if I need some functionality from manager, discuss with the core team and if required contribute some code into their codebase.

File: main.go
Comment (abtinokhovat) [inline main.go:1]: transfer to the cmd folder

File: manager/delivery/httpserver/userhandler/login.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	if result != nil {
+		msg, code := httpmsg.Error(result.Err)
+		if result.Fields["email"] == "user not found" {
+			return ctx.JSON(http.StatusUnauthorized, EchoErrorMessage(errmsg.ErrWrongCredentials))
+		}
+
+		// TODO: in validator we have a ValidatorError struct and this binding is ambiguous, we should change it properly
+		return ctx.JSON(code, echo.Map{
+			"message": msg,
+			"error":   result.Fields,
```
Comment (gocastsian) [inline manager/delivery/httpserver/userhandler/login.go:30]: Change this to `errors`

File: manager/repository/userrepo/user.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package userrepo
+
+import (
+	"fmt"
+
+	"github.com/ormushq/ormus/manager/entity"
+)
+
+func (a StorageAdapter) DoesUserExistsByEmail(email string) (bool, error) {
```
Comment (gocastsian) [inline manager/repository/userrepo/user.go:9]: Move to `repository/scylladb/` and implement

File: manager/service/userservice/register.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+func (s Service) Register(req param.RegisterRequest) (*param.RegisterResponse, error) {
+	hashedPassword, err := password.HashPassword(req.Password)
+	if err != nil {
+		return nil, richerror.New("register.hash").WhitWarpError(err)
+	}
+
+	user := entity.User{
+		CreatedAt: time.Now().UTC(),
```
Comment (gocastsian) [inline manager/service/userservice/register.go:19]: Delegate this to repo.

File: manager/service/userservice/register.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+	}
+
+	createdUser, err := s.repo.Register(user)
+	if err != nil {
+		return nil, richerror.New("register.repo").WhitWarpError(err)
+	}
+
+	// return create new user
+	return &param.RegisterResponse{
+		Email: createdUser.Email,
```
Comment (gocastsian) [inline manager/service/userservice/register.go:34]: Return `user.ID`

File: manager/service/userservice/register.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	}, nil
+}
```
Comment (gocastsian) [inline manager/service/userservice/register.go:36]: publish `user.registered` event.

File: manager/repository/userrepo/user.go (hunk 2 above)
Comment (sinaw369) [reply manager/repository/userrepo/user.go]: moved but not implemented yet

**Comment by abtinokhovat** [comment]
all problems are now resolved @gohossein 
we can merge the PR now .

**Review by gocastsian — APPROVED** [review]
Damet Garm!

**Pull request merged**